codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java

```diff
@@ -217,10 +218,56 @@ protected void deserializeStructure(GenerationContext context, StructureShape sh
     @Override
     protected void deserializeUnion(GenerationContext context, UnionShape shape) {
         GoWriter writer = context.getWriter();
-        // TODO: implement union deserialization
-        // The tricky part is going to be accumulating bytes for unknown members.
-        LOGGER.warning("Union type is currently unsupported for JSON deserialization.");
-        context.getWriter().writeDocs("TODO: implement union serialization.");
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        writeJsonTokenizerStartStub(writer, shape);
+
+        writer.write("var uv $P", symbol);
+
+        writer.write("t, err := decoder.Token()");
+        writer.write("if err != nil { return err }");
+        writer.openBlock("switch t {", "}", () -> {
+            Set<MemberShape> members = new TreeSet<>(shape.members());
+            for (MemberShape member : members) {
+                if (!memberFilter.test(member)) {
+                    continue;
+                }
+
+                Shape target = context.getModel().expectShape(member.getTarget());
+                Symbol memberSymbol = symbolProvider.toSymbol(member);
+                String exportedMemberName = symbol.getName() + symbolProvider.toMemberName(member);
+                String serializedMemberName = getSerializedMemberName(member);
+
+                // This is the function to take a value and convert it to the union type.
+                Symbol conversionFunction = SymbolUtils.createValueSymbolBuilder(
+                        String.format("As%s", exportedMemberName),
+                        symbol.getNamespace()
+                ).build();
+
+                writer.openBlock("case $S:", "", serializedMemberName, () -> {
+                    writer.write("var mv $P", memberSymbol);
+                    target.accept(getMemberDeserVisitor(member, "mv"));
+                    writer.write("uv = $T(mv)", conversionFunction);
+                });
+            }
+
+            writer.openBlock("default:", "", () -> {
+                // This is the function to take a value and convert it to the union type.
+                Symbol conversionFunction = SymbolUtils.createValueSymbolBuilder(
+                        String.format("As%sUnknown", symbol.getName()),
+                        symbol.getNamespace()
+                ).build();
+                writer.addUseImports(AwsGoDependency.AWS_REST_JSON_PROTOCOL);
+                writer.write("tagString, ok := t.(string)");
```

Comment:
Is `len(t) == 0` a valid case for UnknownField?

Reply:
No, a tag can't be empty. It would be impossible to define.